streamcopy/copy.go

```diff
@@ -0,0 +1,54 @@
+package streamcopy
+
+import (
+	"io"
+	"time"
+
+	"github.com/containers/image/types"
+)
+
+const readerSize = 32768
+
+// WithProgress implements io.Copy with a buffered reader, then measures data
+// flow and on the interval -- if still running -- reports offset and artifact
+// information to the signal via the channel.
+func WithProgress(writer *io.PipeWriter, reader io.Reader, artifact types.BlobInfo, interval time.Duration, signal chan types.ProgressProperties) {
+	defer writer.Close()
+
+	count := uint64(0)
+	buf := make([]byte, readerSize)
+
+	t := time.Now()
+
+	for {
+		rn, err := reader.Read(buf)
+		count += uint64(rn)
+
+		if rn > 0 && err == io.EOF {
+			goto write
+		}
+
+		if err != nil {
+			writer.CloseWithError(err)
+			return
+		}
+
+		if interval > 0 && signal != nil {
+			if time.Since(t) > interval {
+				signal <- types.ProgressProperties{Offset: count, Artifact: artifact}
+				t = time.Now()
+			}
```

Comment:
As for the cause of the test failure, _if_ I understand correctly what the test is testing for:

I think the cause is that `reader.Read` above may block for _much_ more than `interval`, but this code only sends one progress report. (This should be possible to verify by adding logging of (timestamp, artifact, offset) to the loop consuming the progress updates.)

To solve this, there could be another goroutine which converts the input `io.Reader` into a channel (of buffers?), and this top-level goroutine would `selectd` on that channel of input data and a `time.Ticker` or something like that.

But that’s much more fairly code; I’d be perfectly fine with keeping ~the current logic and just giving up on the promise that there will be an update about every `interval`.